File: 01026. Maximum Difference Between Node and Ancestor/solution.js
/**
 * Problem: 1026. Maximum Difference Between Node and Ancestor
 *
 * Difficulty: Medium
 *
 * Language: JavaScript
 *
 * Performance: Runtime - 0 ms (Beats 100%)
 */

/**
 * Given a binary tree, find the maximum absolute difference between any ancestor and descendant node values
 *
 * @param {TreeNode} root - Root of the binary tree
 *
 * @returns {number} - Maximum absolute difference between any ancestor and descendant node values
 */
const maxAncestorDiff = (root) => {
  // Initialize the maximum absolute difference
  let absDiff = -1

  // Helper function to traverse the tree and find the maximum absolute difference
  const findDiff = (root) => {
    // Base case: if the node is null, return a sentinel value
    if (root === null) return [-1, Infinity]

    // Recursively find the minimum and maximum values in the left and right subtrees
    const left = findDiff(root.left),
      right = findDiff(root.right)

    // Update the maximum absolute difference using the current node's value
    // and the minimum and maximum values from the left and right subtrees
    const minVal = Math.min(root.val, Math.min(left[1], right[1])),
      maxVal = Math.max(root.val, Math.max(left[0], right[0]))

    // Update the global maximum absolute difference
    absDiff = Math.max(
      absDiff,
      Math.abs(root.val - minVal),
      Math.abs(root.val - maxVal)
    )

    // Return the current node's value as the maximum and minimum for its subtree
    return [maxVal, minVal]
  }

  // Start the recursive traversal from the root
  findDiff(root)

  // Return the maximum absolute difference found
  return absDiff
}

File: 01026. Maximum Difference Between Node and Ancestor/solution.go
/**
 * Problem: 1026. Maximum Difference Between Node and Ancestor
 *
 * Difficulty: Medium
 *
 * Language: Golang
 *
 * Performance: Runtime - 0 ms (Beats 100%)
 */

func maxAncestorDiff(root *TreeNode) int {
	// Initialize the maximum difference found so far to 0
	var maxDiff int

	// Define a recursive helper function that performs depth-first search
	// Parameters: current node, minimum value seen in current path, maximum value seen in current path
	var dfs func(node *TreeNode, minVal, maxVal int)
	
	// Implementation of the DFS function
	dfs = func(node *TreeNode, minVal, maxVal int) {
		// Base case: if current node is nil, return immediately
		if node == nil {
			return
		}

		// Update minimum value if current node's value is smaller
		if node.Val < minVal {
			minVal = node.Val
		}
		// Update maximum value if current node's value is larger
		if node.Val > maxVal {
			maxVal = node.Val
		}

		// Calculate the difference between max and min values in current path
		diff := maxVal - minVal
		// Update global maximum difference if current difference is larger
		if diff > maxDiff {
			maxDiff = diff
		}

		// Recursively traverse left subtree with updated min/max values
		dfs(node.Left, minVal, maxVal)
		// Recursively traverse right subtree with updated min/max values
		dfs(node.Right, minVal, maxVal)
	}

	// Start DFS from root node with initial min and max values set to root's value
	dfs(root, root.Val, root.Val)

	return maxDiff
}
